Stream timings in metadata

kgoodier opened the pull request:
Expose SPDYMetadata as a concrete class, defined in SPDYProtocol.h. No longer are the metadata fields returned in a dictionary as string values.

Add very detailed timings for a stream's lifetime and expose in the metadata. These include the start of a request, end of headers, start of data, end of data, end of a request, then again for a response. Values returned are in seconds and are relative to an arbitrary base (mach_absolute_time). This effectively makes them absolute, so comparing across streams is relatively simple and even comparing time between the app sending a request and CocoaSPDY first receiving it is possible.

kgoodier committed:
Expose metadata as a concrete object.

The metadata is no longer returned as a dictionary of string values to
the parent app. Instead, it is a concrete interface with typed
properties.

kgoodier committed:
Move SPDYMetadataUtils to a category

kgoodier committed:
Make SPDYStopWatch mockable.

Provides a "sleep" method that just applies an offset to the timings.
Only applicable for COVERAGE (unit test) builds. Also expose the start
times and alter the unit tests and clean up the build scheme for unit
tests.

kgoodier committed:
Track detailed stream timings and expose in metadata.

These timings use opaque absolute timestamps. While this makes it a bit
harder to consume, it gives more flexibility and provides more
information. The time between the app starting the request and the
SPDY protocol first receiving it can be measured, for instance. It will
also allow for future additions without breaking the meaning of any
existing timings.

NSProgrammer commented:
-1, don't expose the properties of the metadata as publicly mutable.  You can make them mutable internally to CocoaSPDY, but consumers should not be able to mutate the meta data

kgoodier committed:
Make SPDYMetadata properties publicly read-only

The properties are declared readwrite in a private category for
CocoaSPDY usage.

kgoodier committed:
Simplify SPDYStopwatch for unit tests.

kgoodier committed:
Merge branch 'develop' into feature/detailed-stream-timings

Conflicts:
	SPDY/SPDYProtocol.m
	SPDY/SPDYStream.m
